pricegraph/src/graph/bellman_ford.rs

```diff
@@ -0,0 +1,149 @@
+/// containing the weight of the shortest paths and the predecessor vector for
+/// the shortests paths.
+pub type Paths<G> = (
+    Vec<<G as Data>::EdgeWeight>,
+    Vec<Option<<G as GraphBase>::NodeId>>,
+);
+
+/// A negative cycle error with a path representing the cycle.
+pub struct NegativeCycle<G: Data>(pub Vec<G::NodeId>);
+
+/// This implementation is taken from the `petgraph` crate with a small
+/// modification to return the path when a negative cycle is detected.
+///
+/// The orginal source can be found here:
+/// https://docs.rs/petgraph/0.5.0/src/petgraph/algo/mod.rs.html#745-792
+pub fn search<G>(g: G, source: G::NodeId) -> Result<Paths<G>, NegativeCycle<G>>
+where
+    G: NodeCount + IntoNodeIdentifiers + IntoEdges + NodeIndexable,
+    G::EdgeWeight: FloatMeasure,
+{
+    let mut predecessor = vec![None; g.node_bound()];
+    let mut distance = vec![<_>::infinite(); g.node_bound()];
+
+    let ix = |i| g.to_index(i);
+
+    distance[ix(source)] = <_>::zero();
+    // scan up to |V| - 1 times.
+    for _ in 1..g.node_count() {
+        let mut did_update = false;
+        for i in g.node_identifiers() {
+            for edge in g.edges(i) {
+                let i = edge.source();
+                let j = edge.target();
+                let w = *edge.weight();
+                if distance[ix(i)] + w < distance[ix(j)] {
+                    distance[ix(j)] = distance[ix(i)] + w;
+                    predecessor[ix(j)] = Some(i);
+                    did_update = true;
+                }
+            }
+        }
+        if !did_update {
+            break;
+        }
+    }
+
+    // check for negative weight cycle
+    for i in g.node_identifiers() {
+        for edge in g.edges(i) {
+            let j = edge.target();
+            let w = *edge.weight();
+            if distance[ix(i)] + w < distance[ix(j)] {
+                // NOTE: The following 4 lines are the only modifications made
+                // to the original algorithm and where originally:
+                // ```diff
+                // -//println!("neg cycle, detected from {} to {}, weight={}", i, j, w);
+                // -return Err(NegativeCycle(()));
+                // ```
+                predecessor[ix(j)] = Some(i);
+                let cycle = find_cycle::<G>(g, j, &predecessor)
+                    .expect("negative cycle not found after being detected");
+                return Err(NegativeCycle(cycle));
+            }
+        }
+    }
+
+    Ok((distance, predecessor))
+}
+
+/// Finds a cycle and returns a vector representing a path along the cycle,
+/// ending that is the predecessor of the starting node.
+///
+/// Returns `None` if no such cycle can be found.
+fn find_cycle<G>(
+    g: G,
+    start: G::NodeId,
+    predecessor: &[Option<G::NodeId>],
+) -> Option<Vec<G::NodeId>>
+where
+    G: NodeCount + NodeIndexable,
+{
+    let ix = |i| g.to_index(i);
+
+    // NOTE: First find a node that is actually on the cycle, this is done
```

Comment:
But are we guaranteed to find it by walking backwards? This is what we doing, right?

Reply:
I think so... It is what we are doing.

Reply:
So Bellman-Ford is guaranteed to complete in |V| - 1 iterations, and if it doesn't that means that there is a negative cycle somewhere. They are detected by finding a node that relaxes on the |V|th iteration where each iteration relaxes all adjacent nodes. This means, intuitvely, that a node relaxed on the |V|th iteration must be connected to the negative cycle or it would not be relaxable (since its calculated distance would already be guaranteed to be the lowest). So by walking backwards you are guaranteed to stumble upon the negative cycle.

The question becomes is it possible that the negative cycle does not appear in the `predecessors` vector? That may be possible. It *might* be that the |V|th iteration must be done to completion so that the negative cycle is in `predecessor`s although my intuition says it doesn't and that just finding one node that relaxes on the |V|th cycle is enough to ensure that it is complete in the `predecessors` vector.